plugins/inputs/prometheus_scraper/start.go

```diff
@@ -288,44 +283,29 @@ func reloadConfig(filename string, logger log.Logger, rls ...func(*config.Config
 		return errors.Wrapf(err, "couldn't load configuration (--config.file=%q)", filename)
 	}
 
-	// For saving name before relabel
-	// - __name__ https://github.com/aws/amazon-cloudwatch-agent/issues/190
-	// - job and instance https://github.com/aws/amazon-cloudwatch-agent/issues/193
-	for _, scrapeConfig := range conf.ScrapeConfigs {
-		relabelConfigs := []*relabel.Config{
-			// job
-			{
-				Action:       relabel.Replace,
-				Regex:        relabel.MustNewRegexp(".*"), // __address__ is always there, so we will find a match for every job
-				Replacement:  scrapeConfig.JobName,        // value is hard coded job name
-				SourceLabels: model.LabelNames{"__address__"},
-				TargetLabel:  savedScrapeJobLabel, // creates a new magic label
-			},
-			// instance
-			{
-				Action:       relabel.Replace,
-				Regex:        relabel.MustNewRegexp("(.*)"),
-				Replacement:  "$1", // value is actual __address__, i.e. instance if you don't relabel it.
-				SourceLabels: model.LabelNames{"__address__"},
-				TargetLabel:  savedScrapeInstanceLabel, // creates a new magic label
-			},
+	for _, sc := range conf.ScrapeConfigs {
+		// Add metric relabel to preserve the original name as label and prioritize this metric relabel
+		// before any others metric relabel
+		relabelMetricName := false
+		for _, rc := range sc.MetricRelabelConfigs {
+			if rc.TargetLabel == "__name__" {
+				relabelMetricName = true
+			}
 		}
 
-		// We only got __name__ after scrape, so it's in metric_relabel_configs instead of relabel_configs.
-		metricRelabelConfigs := []*relabel.Config{
-			// __name__
-			{
-				Action:       relabel.Replace,
-				Regex:        relabel.MustNewRegexp("(.*)"),
-				Replacement:  "$1",
-				TargetLabel:  savedScrapeNameLabel,
-				SourceLabels: model.LabelNames{"__name__"},
-			},
+		if relabelMetricName {
```

Comment:
So if any of the metrics have `__name__`, then all of the metrics get relabeled?

Reply:
Yes. To be more exactly, if the customer provides their configuration with target label as __name__, we would an extra label (a magic label) to store the original name before relabeling. OTEL does not allow the customers to [modify the metric name](https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/main/receiver/prometheusreceiver/config.go#L198-L204) and this seems to be an issue in Prometheus too (since prometheus updates the  [metadata cache is only updated before apply relabeling](https://github.com/prometheus/prometheus/blob/e5478983c43cb839a77c144f953b9ea25783b995/scrape/scrape.go#L1280) so we have no control over it)